Emby.Server.Implementations/SyncPlay/SyncPlayManager.cs

```diff
@@ -41,338 +44,415 @@ public class SyncPlayManager : ISyncPlayManager, IDisposable
+            lock (group)
             {
-                // Otherwise show all available groups
-                return _groups.Values.Where(
-                    group => HasAccessToItem(user, group.GetPlayingItemId())).Select(
-                    group => group.GetInfo()).ToList();
+                group.HandleRequest(session, request, cancellationToken);
             }
         }
 
-        /// <inheritdoc />
-        public void HandleRequest(SessionInfo session, PlaybackRequest request, CancellationToken cancellationToken)
+        /// <summary>
+        /// Releases unmanaged and optionally managed resources.
+        /// </summary>
+        /// <param name="disposing"><c>true</c> to release both managed and unmanaged resources; <c>false</c> to release only unmanaged resources.</param>
+        protected virtual void Dispose(bool disposing)
         {
-            var user = _userManager.GetUserById(session.UserId);
-
-            if (user.SyncPlayAccess == SyncPlayAccess.None)
+            if (_disposed)
             {
-                _logger.LogWarning("HandleRequest: {0} does not have access to SyncPlay.", session.Id);
+                return;
+            }
 
-                var error = new GroupUpdate<string>()
-                {
-                    Type = GroupUpdateType.JoinGroupDenied
-                };
+            _sessionManager.SessionStarted -= OnSessionManagerSessionStarted;
+            _disposed = true;
+        }
 
-                _sessionManager.SendSyncPlayGroupUpdate(session.Id, error, CancellationToken.None);
-                return;
+        private void OnSessionManagerSessionStarted(object sender, SessionEventArgs e)
+        {
+            var session = e.SessionInfo;
+
+            Guid groupId = Guid.Empty;
+            lock (_mapsLock)
+            {
+                groupId = FindJoinedGroupId(session);
             }
 
-            lock (_groupsLock)
+            if (groupId.Equals(Guid.Empty))
             {
-                _sessionToGroupMap.TryGetValue(session.Id, out var group);
+                return;
+            }
 
-                if (group == null)
-                {
-                    _logger.LogWarning("HandleRequest: {0} does not belong to any group.", session.Id);
+            var request = new JoinGroupRequest(groupId);
+            JoinGroup(session, request, CancellationToken.None);
+        }
 
-                    var error = new GroupUpdate<string>()
-                    {
-                        Type = GroupUpdateType.NotInGroup
-                    };
-                    _sessionManager.SendSyncPlayGroupUpdate(session.Id, error, CancellationToken.None);
-                    return;
-                }
+        /// <summary>
+        /// Checks if a given session has joined a group.
+        /// </summary>
+        /// <remarks>
+        /// Method is not thread-safe, external locking on <see cref="_mapsLock"/> is required.
```

Comment:
As long as a request is processed singularly for a certain group all should be fine. And the locks on `IGroupController` try to guarantee this. Clients can send updates at their own pace, what matters is that an action is isolated from others pertaining the same group.

The `_groupsLock` and `_mapsLock` locks, on the other hand, control the access flow to `_groups` and `_sessionToGroupMap`, respectively, and these two containers are not thread-safe. The methods that handle them inherit this non thread-safety.

In other words, requests are concurrent between groups and only if they don't make changes to the list of groups/participants.

Finally, to avoid deadlocks, the locks are acquired and released following the same order.

Reply:
I would need to spend more time looking through the code on my local machine so pardon my asking, but could we not use the Concurrent collections and employ optimistic locking? I assume that individual members can only manipulate their own state and that a potential admin can manipulate them all. But if someone is kicked or leaves the room, then the locking isn't strictly necessary as the 2nd removal will just be a noop.

Which actions require pessimistic locking to not cause unwanted race conditions?

Reply:
I must confess that optimistic locking is one of those obvious solutions/approaches one goes for, if only I wasn't completely oblivious to it. I wasn't aware that the concurrent version of the collections work this way.

> Which actions require pessimistic locking to not cause unwanted race conditions?

I think that the problematic ones are those related to joining/leaving groups.
- Joining a group can potentially change another group's state as well: a user, that has joined a group, wants to join another group without first explicitly leaving the current one. The server allows this and handles the leave-old-group/join-new-group steps in the same request.
- Same goes for creating a new group while being in another one.
- Another edge case that needs to be handled is the following: [last member is leaving a group, which is now about to be destroyed, and another member attempts to join this same group](https://github.com/jellyfin/jellyfin/pull/2733#discussion_r411714459).

I'll try bringing back the concurrent collections and drop one of the two locks in `SyncPlayManager`, probably `_mapsLock`.

Reply:
> The server allows this and handles the leave-old-group/join-new-group steps in the same request.

Dropping this would simplify things a lot. :thinking:

Reply:
> Another edge case that needs to be handled is the following: last member is leaving a group, which is now about to be destroyed, and another member attempts to join this same group.

Valid point and we should indeed be careful when performing actions that affect the others in the group, but I just think the current locking scheme is a little aggressive.

The thing about Concurrent collections is that the guarantee thread safety, _but not_ necessarily safety of the values. Two threads can overwrite each other if you're not careful. That's the pitfall. I do think there's a reasonable middle ground hidden somewhere here though.

Reply:
Noticed another "bug" that can happen in the following way:
- A `HandleRequest` get the reference to a Group but does not manage to lock on it.
- A `LeaveGroup` request gets a reference to the same Group and manages to lock it; Group is now empty and will be removed.
- `HandleRequest` now locks on an empty Group.

Nothing too bad per se should happen, but it's not that right either.
I think that I'm going to add a property to `Group` to signal whether it's still active otherwise the requests are dropped (probably checking if it's empty is enough).

Reply:
Done. The code should be cleaner now.